torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1606,49 +1606,37 @@ IterDomain* IterDomain::merge(IterDomain* outer, IterDomain* inner) {
   TORCH_CHECK(
       !outer->extent()->isZeroInt() && !inner->extent()->isZeroInt(),
       "Merging IterDomains with ending values that are 0 is not supported at this time.");
-  TORCH_CHECK(
-      outer->isReduction() == inner->isReduction() ||
-          (!outer->isReduction() && inner->isTrivialReduction()) ||
-          (outer->isTrivialReduction() && !inner->isReduction()),
-      "Merging IterDomains requires that their iteration types match.");
-  TORCH_CHECK(
-      (outer->isGather() && inner->isGather()) ||
-          (!outer->isGather() && !inner->isGather()),
-      "Merging gather and non-gather domains is not supported.");
-
-  TORCH_CHECK(
-      !outer->isStride() && !inner->isStride(),
-      "No support for merging stride domains");
 
   Val* merged_id_size = mul(outer->extent(), inner->extent());
 
   IterType itype = outer->getIterType();
 
-  if (outer->isBroadcast() && inner->isBroadcast()) {
-    itype = IterType::Broadcast;
+  if (inner->getIterType() == itype) {
```

Comment:
Are there any logical changes here? Or does the new version implement the same logic just in a different way? Looks like it's the latter, but if that's the case, it seems to me the previous version is easier to understand. It's more verbose, but it makes the conditions explicit, which I think really helpful to understand the underlying logic.

Reply:
They are not the same logic. The previous logic does not allow me to merge a (non-trivial) reduction with a broadcasting. And what I am trying to implement here is a pattern matching. I am trying to match the following rules in order. If one rule matches, then we don't proceed with later rules.

```
X + X = X
trivial reduction + X = X
X + trivial reduction = X
broadcasting + X = X
X + broadcasting = X
fail
```
I think this rule could contain all the previous checks, and if-goto seems to be the most natural way to implement this pattern matching in my mind.

If I explain this idea in a comment, something like
```C++
// The itertype inference is a pattern matching of the rules below:
//
//   X + X = X
//   trivial reduction + X = X
//   X + trivial reduction = X
//   broadcasting + X = X
//   X + broadcasting = X
//   fail
//
// The rules are proceeded one by one in order. For each rule, we test if the
// given (outer, inner) matches the pattern. If it does, then we stop
// procceeding and get a result. If we have reached the end without finding any
// matched pattern, then it is a mistake and should be reported.
```
Would you think it will be clear enough, or do you think it is still better to go back to the previous way?

Reply:
OK, I didn't notice merging of reductions and broadcasts. Please add the comments to the code.

I don't agree with the formulation with `goto`. Unless there's no other reasonable way, `goto` shouldn't be used. I think we could easily rewrite this logic with `if` and `else if`, or we could have a separate function that just returns `IterType`, which basically we would just need to replace `goto` with `return`.

Reply:
Also, I think it's worth mentioning that `reduction + broadcast => reduction`, while `trivial reduction + broadcast => broadcast`. The former is obvious as we can't ignore the reduction unless it's trivial. The latter could yield a `trivial reduction`, which might be more consistent. I suppose it doesn't really matter if it's a trivial reduction or a broadcast, so the current logic would be fine, but would be helpful to have a comment on it.

Reply:
fixed